lib/internal/filter.js

```diff
@@ -1,31 +1,27 @@
-import arrayMap from 'lodash/_arrayMap';
-import property from 'lodash/_baseProperty';
+import isArrayLike from 'lodash/isArrayLike';
 import noop from 'lodash/noop';
+
 import once from './once';
+import iterator from './iterator';
 
-export default function _filter(eachfn, arr, iteratee, callback) {
+export default function _filter(eachfn, coll, iteratee, callback) {
     callback = once(callback || noop);
-    var results = [];
-    eachfn(arr, function (x, index, callback) {
+    var truthValues = isArrayLike(coll) ? new Array(coll.length) : {};
+    eachfn(coll, function (x, index, callback) {
         iteratee(x, function (err, v) {
-            if (err) {
-                callback(err);
-            }
-            else {
-                if (v) {
-                    results.push({index: index, value: x});
-                }
-                callback();
-            }
+            truthValues[index] = !!v;
+            callback(err);
         });
     }, function (err) {
-        if (err) {
-            callback(err);
-        }
-        else {
-            callback(null, arrayMap(results.sort(function (a, b) {
-                return a.index - b.index;
-            }), property('value')));
+        if (err) return callback(err);
+        var result = [];
+        var nextElem = iterator(coll);
+        var elem;
+        while ((elem = nextElem()) !== null) {
```

Comment:
So there is a drawback/regression for doing it this way:

When `coll` is an iterator we will be trying to instantiate the iterator twice; sometimes an iterator can only be run once which would cause us to return an empty list.